pkg/gofr/datasource/solr/solr_test.go

```diff
@@ -0,0 +1,168 @@
+package solr
+
+import (
+	"bytes"
+	"context"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"go.uber.org/mock/gomock"
+)
+
+func Test_InvalidRequest(t *testing.T) {
+	_, err := call(context.TODO(), "GET", ":/localhost:", nil, nil)
+
+	require.Error(t, err, "TEST Failed.\n")
+}
+
+func Test_InvalidJSONBody(t *testing.T) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		_, _ = w.Write([]byte(`Not a JSON`))
+	}))
+	defer ts.Close()
+
+	_, err := call(context.TODO(), "GET", ts.URL, nil, nil)
+
+	require.Error(t, err, "TEST Failed.\n")
+}
+
+func TestSolr(t *testing.T) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		_, _ = w.Write([]byte(`{
+  		"responseHeader": {
+    	"rf": 1,
+    	"status": 0}}`))
+	}))
+
+	defer ts.Close()
+
+	a := ts.Listener.Addr().String()
+	addr := strings.Split(a, ":")
+
+	ctrl := gomock.NewController(t)
+	mockLogger := NewMockLogger(ctrl)
+	mockMetrics := NewMockMetrics(ctrl)
+
+	mockLogger.EXPECT().Debug(gomock.Any()).Times(9)
+	mockMetrics.EXPECT().RecordHistogram(gomock.Any(), "app_solr_stats", gomock.Any(), "type", gomock.Any()).Times(9)
+
+	s := New(Config{Host: addr[0], Port: addr[1]})
+	s.metrics = mockMetrics
+	s.logger = mockLogger
+
+	testClientSearch(t, s)
+	testClientAddField(t, s)
+	testClientCreate(t, s)
+	testClientDelete(t, s)
+	testClientDeleteField(t, s)
+	testClientListFields(t, s)
+	testClientRetrieve(t, s)
+	testClientUpdate(t, s)
+	testClientUpdateField(t, s)
+}
+
+func testClientSearch(t *testing.T, s *Client) {
```

Comment:
This is a strange way to do

I would add `setupClient(t *testing)` that does everything before you call the test client

Then I would add a TestClientSearch that calls setupClient and get `s` from it